app/Domain/Auth/Controllers/CallbackController.php

```diff
@@ -20,13 +21,37 @@ public function __invoke(Request $request): RedirectResponse
 
         if ($auth0->exchange()) {
             $user = $auth0->getUser();
-            Session::put('id_token', $auth0->getIdToken());
 
-            if ($user !== null) {
-                Auth::login(UserModel::fromSession($user));
+            if ($user === null) {
+                return redirect()->intended('/');
+            }
+
+            if (config(KeycloakConfig::KEYCLOAK_ENFORCE_2FA_FOR_ADMINS)) {
+                $user['acr'] = $user['acr'] ?? null;
+                if ($user['acr'] !== 'highest' && in_array($user['email'], config('nova.users'), true)) {
+                    // Calling clear() is necessary here. Without it, the user will be  already logged in on the frontend but doesn’t have the correct ACR level for the admin, they get stuck in a redirect loop — they appear logged in, but don't meet the access requirements, so they keep getting kicked back without triggering a proper re-login with higher privileges.
+                    $auth0->clear();
+
+                    $url = $auth0->login(null, $this->addAcrEnforcementParam());
+
+                    return redirect()->to($url);
+                }
             }
+
+            Session::put('id_token', $auth0->getIdToken());
+
+            Auth::login(UserModel::fromSession($user));
         }
 
         return redirect()->intended('/');
     }
+
+    public function addAcrEnforcementParam(): array
+    {
+        $params = [];
+        parse_str(config(KeycloakConfig::KEYCLOAK_LOGIN_PARAMETERS), $params);
```

Comment:
Inject the config, this will make the tests easier to maintain. You can check the bootstrapping of the `LoginController` to see how to bind the config values to the input parameter of a constructor.